contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.2.0--3.3.0.sql

```diff
@@ -424,6 +424,375 @@ END
 $$ LANGUAGE plpgsql;
 GRANT EXECUTE ON PROCEDURE sys.bbf_sleep_until(IN sleep_time DATETIME) TO PUBLIC;
 
+-- sp_babelfish_autoformat is a helper procedure which formats the contents of a table (or view)
+-- as narrowly as possible given its actual column contents.
+-- This proc is currently only used by sp_who but could be applied more generically.
+-- A complication is that the metadata from #tmp tables cannot be found in the babelfish
+-- catalogs, so we have to use some trickery to make things work.
+-- Not all datatypes are handled as well as might be possible, but it is sufficient for 
+-- the current purposes.
+CREATE OR REPLACE PROCEDURE sys.sp_babelfish_autoformat(
+	IN "@tab"        VARCHAR(257) DEFAULT NULL,
+	IN "@orderby"    VARCHAR(1000) DEFAULT '',
+	IN "@printrc"    sys.bit DEFAULT 1,
+	IN "@hiddencols" VARCHAR(1000) DEFAULT NULL)
+LANGUAGE 'pltsql'
+AS $$
+BEGIN
+	SET NOCOUNT ON
+	DECLARE @rc INT
+	DECLARE @id INT
+	DECLARE @objtype VARCHAR(2)	
+	DECLARE @msg VARCHAR(200)	
+	
+	IF @tab IS NULL
+	BEGIN
+		RAISERROR('Must specify table name', 16, 1)
+		RETURN		
+	END
+	
+	IF TRIM(@tab) = ''
+	BEGIN
+		RAISERROR('Must specify table name', 16, 1)
+		RETURN		
+	END	
+	
+	-- Since we cannot find #tmp tables in the Babelfish catalogs, we cannot check 
+	-- their existence other than by trying to select from them
+	-- NB: not handling uncommon but valid syntax '<schemaname>.#tmp' for #tmp tables
+	IF SUBSTRING(@tab,1,1) <> '#'
+	BEGIN
+		SET @id = OBJECT_ID(@tab)
+		IF @id IS NULL
+		BEGIN
+			IF SUBSTRING(UPPER(@tab),1,4) = 'DBO.'
+			BEGIN
+				SET @id = OBJECT_ID('SYS.' + SUBSTRING(@tab,5))
+			END
+			IF @id IS NULL
+			BEGIN		
+				SET @msg = 'Table or view '''+@tab+''' not found'
+				RAISERROR(@msg, 16, 1)
+				RETURN		
+			END
+		END
+	END
+	
+	SELECT @objtype = type FROM sys.sysobjects WHERE id = @id
+	IF @objtype NOT IN  ('U', 'S', 'V')
+	BEGIN
+		SET @msg = ''''+@tab+''' is not a table or view'
+		RAISERROR(@msg, 16, 1)
+		RETURN		
+	END
+	
+	-- check for 'ORDER BY', if specified
+	SET @orderby = TRIM(@orderby)
+	IF @orderby <> ''
+	BEGIN
+		IF UPPER(@orderby) NOT LIKE 'ORDER BY%'
+		BEGIN
+			RAISERROR('@orderby parameter must start with ''ORDER BY''', 16, 1)
+			RETURN
+		END
+	END
+	
+	-- columns to hide in final client output
+	-- assuming delimited column names do not contain spaces or commas inside the name
+	WHILE (CHARINDEX(' ,', @hiddencols) > 0) or (CHARINDEX(', ', @hiddencols) > 0)
+	BEGIN
+		SET @hiddencols = UPPER(REPLACE(@hiddencols, ' ,', ','))
+		SET @hiddencols = UPPER(REPLACE(@hiddencols, ', ', ','))
+	END
+	IF LEN(@hiddencols) IS NOT NULL SET @hiddencols = ',' + @hiddencols + ','
```

Comment:
1: TRIM() won't do the job: we're removing an arbitrary number of spaces inside a string here, either before or after a comma. This has to use a loop.
2: UPPER() is used just for defensive coding. It could indeed be moved outside the loop.

Reply:
Defensive coding might lead to performance impact. Taking the UPPER out of the loop might be a better choice.